### PR TITLE
[JENKINS-39867] Add environment variables to container from a secret.

### PR DESCRIPTION
**Pull request by tkgregory**
Previously, environment variables could only be configured as key/value pairs whose value was defined in Jenkins configuration.

This pull request adds the ability to configure an environment variable whose value is derived from a Kubernetes secret. The changes are summarised below:

* Configure UI so that you can add either a `simple` or `secret` environment variable, both for pod and container templates.
* Change pod template configuration class hierarchy to include PodSecretEnvVar, AbstractPodEnvVar, and AbstractEnvVar. PodEnvVar has not been renamed in order to maintain backwards compatibility with previous versions.
* Same changes as above for container template configuration class hierarchy.
* KubernetesCloud now also builds the secret env vars, via a new `buildEnvVar` method. 
* Functional tests and unit tests to cover all scenarios.

### CODE REVIEW

**Comment by carlossg** [comment]
Thanks, I see a couple problems
* public methods being removed, breaking binary compatibility 
* empty space/imports changes that make it difficult to review and can change merge conflicts

**Comment by tkgregory** [comment]
Thanks @carlossg. 

- I have removed the whitespace and the rearrangement of the imports
- Which public methods are you referring to? It may appear as though the `getKey` method has been removed from `ContainerEnvVar` and `PodEnvVar` but they have actually just moved up to a superclass `AbstractEnvVar`.

**Comment by carlossg** [comment]
needs rebase :\

**Comment by tkgregory** [comment]
@carlossg Has been rebased now.

**Comment by tkgregory** [comment]
@carlossg Any chance you could merge this now as I won't be able to work on this for much longer. Thanks.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java [hunk 1]
```diff
@@ -287,4 +267,19 @@ public static String substitute(String s, Map<String, String> properties, String
         }
         return s;
     }
+
+    private static List<AbstractContainerEnvVar> combineEnvVars(ContainerTemplate parent, ContainerTemplate template) {
+        List<AbstractContainerEnvVar> combinedSimpleEnvVars = new ArrayList<>();
```
Comment (iocanel) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java]: I think that the name here is misleading, right?

It seems that we are talking about all env vars not just simple, right?

**Comment by iocanel** [comment]
@carlossg: Do you mind bringing this in? As you initially requested changes, I'd like to hear you opinion before merging it.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java (hunk 1 above)
Comment (tkgregory) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java]: You are right, it is both types. I have renamed the variables accordingly.

**Comment by carlossg** [comment]
I have a working branch with these changes and some more tests, will push in a bit

**Comment by carlossg** [comment]
I have removed the duplication betwen container and pod envVars and added pipeline tests, leaving `envVar` for all key/value env vars and `secretEnvVar` for all coming from a secret.

The only problem right now is that the test fails with `Undefined symbol 'envVar'` and haven't found yet the cause

    mvn clean test -Dtest=KubernetesPipelineTest#runWithEnvVariables

**Comment by carlossg** [comment]
This is now ready to be merged

**Comment by MattLud** [comment]
Can someone confirm status of this merge? 

I am seeing two issues when using master with this branch

- Error on System Configuration page when visiting after upgrade; saving removes existing environment variables
- Unable to access dropdown menu on System Variables; still occurred when using newer code from #197 as well.

**Comment by carlossg** [comment]
@mattlud can you provide an XML config file that exhibits than problems after upgrade.
Also what version of master are you using?

**Comment by carlossg** [comment]
You should be using master, not this branch, due to #197

**Comment by MattLud** [comment]
I was using latest master when testing with 2.70 and 2.32.1(default pom.xml). 

I can get the config.xml on Monday.

**Comment by MattLud** [comment]
Looks like the existing variables are no longer deleted. 

https://gist.github.com/MattLud/cec8a3a7e63c89beb88f7c59fd8aef4e

Version of code being used rev: 5d4bb7 (today's master) 

Steps to recreate: 

1. Checkout 9989fff in fresh workspace
2. Update credentials plugin
3. Add cloud and then go add a template with a test environment variable. 
4. Save and stop the jenkins server. 
5. Update to latest master & start server
6. Go back to configuration, notice the drop down for pod system variables no longer works.

**Comment by MattLud** [comment]
Also noticed this in my logs 

WARNING: Father of PodEnvVar [getValue()=sample, getKey()=sample] and its getDescriptor() points to two different instances. Probably malplaced @Extension. See http://hudson.361315.n4.nabble.com/Help-Hint-needed-Post-build-action-doesn-t-stay-activated-td2308833.html

**Pull request merged**